src/crypto.rs

```diff
@@ -1,83 +1,59 @@
-use std::sync::Arc;
-
 use base64;
 use ring::constant_time::verify_slices_are_equal;
 use ring::{digest, hmac, rand, signature};
-use std::str::FromStr;
 use untrusted;
 
-use errors::{new_error, Error, ErrorKind, Result};
-
-/// The algorithms supported for signing/verifying
-#[derive(Debug, PartialEq, Copy, Clone, Serialize, Deserialize)]
-pub enum Algorithm {
-    /// HMAC using SHA-256
-    HS256,
-    /// HMAC using SHA-384
-    HS384,
-    /// HMAC using SHA-512
-    HS512,
-
-    /// ECDSA using SHA-256
-    ES256,
-
-    /// ECDSA using SHA-384
-    ES384,
-
-    /// RSASSA-PKCS1-v1_5 using SHA-256
-    RS256,
-    /// RSASSA-PKCS1-v1_5 using SHA-384
-    RS384,
-    /// RSASSA-PKCS1-v1_5 using SHA-512
-    RS512,
-}
-
-impl Default for Algorithm {
-    fn default() -> Self {
-        Algorithm::HS256
-    }
-}
-
-impl FromStr for Algorithm {
-    type Err = Error;
-    fn from_str(s: &str) -> Result<Self> {
-        match s {
-            "HS256" => Ok(Algorithm::HS256),
-            "HS384" => Ok(Algorithm::HS384),
-            "HS512" => Ok(Algorithm::HS512),
-            "ES256" => Ok(Algorithm::ES256),
-            "ES384" => Ok(Algorithm::ES384),
-            "RS256" => Ok(Algorithm::RS256),
-            "RS384" => Ok(Algorithm::RS384),
-            "RS512" => Ok(Algorithm::RS512),
-            _ => Err(new_error(ErrorKind::InvalidAlgorithmName)),
-        }
-    }
-}
+use crate::algorithms::Algorithm;
+use crate::errors::{new_error, ErrorKind, Result};
+use crate::keys::Key;
 
 /// The actual HS signing + encoding
-fn sign_hmac(alg: &'static digest::Algorithm, key: &[u8], signing_input: &str) -> Result<String> {
-    let signing_key = hmac::SigningKey::new(alg, key);
+fn sign_hmac(alg: &'static digest::Algorithm, key: Key, signing_input: &str) -> Result<String> {
+    let signing_key = match key {
+        Key::Hmac(bytes) => hmac::SigningKey::new(alg, bytes),
+        _ => return Err(ErrorKind::InvalidKeyFormat)?,
+    };
     let digest = hmac::sign(&signing_key, signing_input.as_bytes());
 
     Ok(base64::encode_config::<hmac::Signature>(&digest, base64::URL_SAFE_NO_PAD))
 }
 
 /// The actual ECDSA signing + encoding
-fn sign_ecdsa(alg: &'static signature::EcdsaSigningAlgorithm, key: &[u8], signing_input: &str) -> Result<String> {
-    let signing_key = signature::EcdsaKeyPair::from_pkcs8(alg, untrusted::Input::from(key))?;
+fn sign_ecdsa(
+    alg: &'static signature::EcdsaSigningAlgorithm,
+    key: Key,
+    signing_input: &str,
+) -> Result<String> {
+    let signing_key = match key {
+        Key::Pkcs8(bytes) => {
+            signature::EcdsaKeyPair::from_pkcs8(alg, untrusted::Input::from(bytes))?
```

Comment:
In *ring* 0.16, all the use of `untrusted` in the public API was removed. Often you can just use the raw byte slice now.